hera_qm/vis_metrics.py

```diff
@@ -48,6 +50,105 @@ def check_noise_variance(data):
     return Cij
 
 
+def sequential_diff(data, wgts=None, axis=(0,)):
+    """
+    Take a sequential (adjacent) difference of a visibility waterfall
+    as an estimate of the visibility noise.
+
+    Parameters
+    ----------
+    data : ndarray or UVData object
+        A 2D or 3D ndarray containing visibility data, with
+        shape (Ntimes, Nfreqs, :). Or UVData object
+        holding visibility data.
+
+    wgts : ndarray
+        A 2D or 3D ndarray containing visibility weights
+        matching data shape. If None, and data is UVData,
+        will use UVData ~flags * nsample as weights.
+
+    axis : int or tuple
+        Axes along which to take sequential difference
+
+    Returns
+    -------
+    diff_data : ndarray or UVData object
+        A 2D or 3D complex differenced data array,
+        or UVData object holding differenced data
+        divided by sqrt(2).
+
+    prod_wgts : ndarray
+        If input data is ndarray, this is the geometric
+        mean of all wgts in the difference.
+    """
+    # type check
+    if isinstance(axis, (int, np.integer)):
+        axis = (axis,)
+    assert np.all([ax in [0, 1] for ax in axis]), "axis must be 0, 1 or both"
+
+    # perform differencing if data is an ndarray
+    if isinstance(data, np.ndarray):
+        # get weights
+        if wgts is None:
+            wgts = np.ones_like(data, dtype=np.float64)
+        else:
+            assert isinstance(wgts, np.ndarray), \
+                   "wgts must be ndarray if data is ndarray"
+
+        # difference data
+        for ax in axis:
+            data = (utils.dynamic_slice(data, slice(1, None), axis=ax) \
+                   - utils.dynamic_slice(data, slice(None, -1), axis=ax)) / np.sqrt(2)
+            wgts = np.sqrt(utils.dynamic_slice(wgts, slice(1, None), axis=ax) \
+                   * utils.dynamic_slice(wgts, slice(None, -1), axis=ax))
```

Comment:
Why is it that weights are multiplied? Eventually this gets reinterpreted as flags and nsamples. I can see how this gets you the correct flagging (if any component is flagged, the result is flagged), but it doesn't seem like the right thing for nsamples.

Reply:
yeah it was done to get the correct flagging patterns, but you are right that its not strictly correct for `nsample`. In fact, upon reflection, I think `wgts` is a bad name b/c we aren't doing any kind of weighting anyways. I'll change this